librubyfmt/src/render_targets.rs

```diff
@@ -150,43 +159,267 @@ impl AbstractTokenTarget for BreakableEntry {
             || self.contains_hard_newline()
     }
 
-    fn len(&self) -> usize {
-        self.tokens.len()
+    fn tokens(&self) -> &Vec<AbstractLineToken> {
+        &self.tokens
+    }
+
+    fn any_collapsing_newline_has_heredoc_content(&self) -> bool {
+        self.tokens().iter().any(|t| match t {
+            AbstractLineToken::CollapsingNewLine(Some(..)) => true,
+            AbstractLineToken::SoftNewline(Some(..)) => true,
+            AbstractLineToken::BreakableEntry(be) => {
+                be.any_collapsing_newline_has_heredoc_content()
+            }
+            _ => false,
+        })
     }
 }
 
 impl BreakableEntry {
-    pub fn new(spaces: ColNumber, delims: BreakableDelims, context: FormattingContext) -> Self {
+    pub fn new(delims: BreakableDelims, context: Vec<FormattingContext>) -> Self {
         BreakableEntry {
-            spaces,
             tokens: Vec::new(),
             line_numbers: HashSet::new(),
             delims,
             context,
         }
     }
 
-    pub fn entry_formatting_context(&self) -> FormattingContext {
+    pub fn in_string_embexpr(&self) -> bool {
         self.context
+            .iter()
+            .any(|fc| fc == &FormattingContext::StringEmbexpr)
+    }
+
+    fn contains_hard_newline(&self) -> bool {
+        self.tokens.iter().any(|t| {
+            matches!(
+                t,
+                AbstractLineToken::ConcreteLineToken(ConcreteLineToken::HardNewLine)
+            )
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct BreakableCallChainEntry {
+    tokens: Vec<AbstractLineToken>,
+    line_numbers: HashSet<LineNumber>,
+    call_chain: Vec<CallChainElement>,
+    context: Vec<FormattingContext>,
+}
+
+impl AbstractTokenTarget for BreakableCallChainEntry {
+    fn to_breakable_entry(self: Box<Self>) -> BreakableEntry {
+        unreachable!()
```

Comment:
yeah I feel very strongly this should option, a programming mistake can cause us to reach an unreachable, which is what option unwrap is for